idaes/core/util/tests/test_model_diagnostics.py

```diff
@@ -0,0 +1,249 @@
+    
+    # Specify Ipopt as the solver
+    opt = pyo.SolverFactory('ipopt')
+    
+    # Specifying an iteration limit of 0 allows us to inspect the initial point
+    opt.options['max_iter'] = 0
+
+    # "Solving" the model with an iteration limit of 0 load the initial point and applies
+    # any preprocessors (e.g., enforces bounds)
+    opt.solve(m2, tee=True)
+    
+    # Create Degeneracy Hunter object
+    dh2 = DegeneracyHunter(m2)
+    
+    # Check for violated constraints at the initial point
+    initial_point_constraints = dh2.check_residuals(tol=0.1)
+    
+    # Check there are 1 constraints with large residuals
+    assert len(initial_point_constraints) == 1
+    
+    initial_point_constraint_names = extract_constraint_names(initial_point_constraints)
+    
+    # Check first constraint
+    assert initial_point_constraint_names[0] == 'con2'
+    
+    # Resolve
+    opt.options['max_iter'] = 500
+    opt.solve(m2, tee=True)
+    
+    # Check solution
+    x_sln = []
+    
+    for i in m2.I:
+        x_sln.append(m2.x[i]())
+    
+    assert pytest.approx(x_sln[0], abs=1E-6) == 1.0
+    assert pytest.approx(x_sln[1], abs=1E-6) == 0.0
+    assert pytest.approx(x_sln[2], abs=1E-6) == 0.0
+    
+
+
+
+# Problem 2 with degenerate constraint
+@pytest.mark.unit
+def test_problem2_with_degenerate_constraint():
+
+    # Create test problem instance
+    m2 = example2(with_degenerate_constraint=True)
+    
+    # Specify Ipopt as the solver
+    opt = pyo.SolverFactory('ipopt')
+
+    # Specifying an iteration limit of 0 allows us to inspect the initial point
+    opt.options['max_iter'] = 0
+
+    # "Solving" the model with an iteration limit of 0 load the initial point and applies
+    # any preprocessors (e.g., enforces bounds)
+    opt.solve(m2, tee=True)
+    
+    # Create Degeneracy Hunter object
+    dh2 = DegeneracyHunter(m2)
+    
+    # Check for violated constraints at the initial point
+    initial_point_constraints = dh2.check_residuals(tol=0.1)
+    
+    # Check there are 2 constraints with large residuals
+    assert len(initial_point_constraints) == 2
+    
+    initial_point_constraint_names = extract_constraint_names(initial_point_constraints)
+    
+    # Check first constraint
+    assert initial_point_constraint_names[0] == 'con2'
+    
+    # Check first constraint
+    assert initial_point_constraint_names[1] == 'con5'
+    
+    # Resolve
+    opt.options['max_iter'] = 500
+    opt.solve(m2, tee=True)
+    
+    # Check solution
+    x_sln = []
+    
+    for i in m2.I:
+        x_sln.append(m2.x[i]())
+    
+    assert pytest.approx(x_sln[0], abs=1E-6) == 1.0
+    assert pytest.approx(x_sln[1], abs=1E-6) == 0.0
+    assert pytest.approx(x_sln[2], abs=1E-6) == 0.0
+    
+    # Check the rank
+    n_rank_deficient = dh2.check_rank_equality_constraints()
+    
+    assert n_rank_deficient == 1
+    
+    # TODO: Add MILP solver to idaes get-extensions and add more tests
```

Comment:
I agree that this is needed. It would also be nice to test the printed/logged output of some of the functions. Pytest captures output to the logger, stdout and stderr which can be used to check for expected output strings.

Reply:
This is a great idea. I added it to the TODO list.